File: internal/replay/volatile.go
package replay

import (
	"encoding/json"
	"reflect"
	"regexp"
	"strings"
)

type VolatileConfig struct {
	IgnoreFields   []string
	IgnorePatterns []*regexp.Regexp
}

func DefaultVolatileConfig() *VolatileConfig {
	return &VolatileConfig{
		IgnoreFields: []string{
			"timestamp",
			"createdAt",
			"updatedAt",
			"id",
			"uuid",
			"requestId",
			"traceId",
			"spanId",
			"date",
			"time",
			"version",
		},
		IgnorePatterns: []*regexp.Regexp{
			regexp.MustCompile(`(?i).*_at$`),
			regexp.MustCompile(`(?i).*_id$`),
			regexp.MustCompile(`(?i).*timestamp.*`),
			regexp.MustCompile(`(?i).*uuid.*`),
		},
	}
}

func normalizeToInterface(jsonStr string, config *VolatileConfig) (any, error) {
	if config == nil {
		var v any
		if err := json.Unmarshal([]byte(jsonStr), &v); err != nil {
			return nil, err
		}
		return v, nil
	}

	var data any
	if err := json.Unmarshal([]byte(jsonStr), &data); err != nil {
		return nil, err
	}

	normalized := removeVolatileFields(data, config)
	return normalized, nil
}

func NormalizeJSON(jsonStr string, config *VolatileConfig) (string, error) {
	normalizedIface, err := normalizeToInterface(jsonStr, config)
	if err != nil {
		return jsonStr, err
	}

	result, err := json.Marshal(normalizedIface)
	if err != nil {
		return jsonStr, err
	}

	return string(result), nil
}

func removeVolatileFields(data any, config *VolatileConfig) any {
	switch v := data.(type) {
	case map[string]any:
		result := make(map[string]any)

		for key, value := range v {
			if shouldIgnoreField(key, config) {
				continue
			}

			result[key] = removeVolatileFields(value, config)
		}

		return result

	case []any:
		result := make([]any, len(v))
		for i, item := range v {
			result[i] = removeVolatileFields(item, config)
		}

		return result

	default:
		return v
	}
}

func shouldIgnoreField(fieldName string, config *VolatileConfig) bool {
	if config == nil {
		return false
	}

	for _, ignore := range config.IgnoreFields {
		if strings.EqualFold(fieldName, ignore) {
			return true
		}
	}

	for _, pattern := range config.IgnorePatterns {
		if pattern.MatchString(fieldName) {
			return true
		}
	}

	return false
}

func CompareWithVolatility(body1, body2 string, config *VolatileConfig) (bool, error) {
	i1, err := normalizeToInterface(body1, config)
	if err != nil {
		return false, err
	}

	i2, err := normalizeToInterface(body2, config)
	if err != nil {
		return false, err
	}

	return reflect.DeepEqual(i1, i2), nil
}

type VolatileDiff struct {
	HasDiff          bool
	VolatileOnly     bool
	StableFieldsDiff bool
	NormalizedBody1  string
	NormalizedBody2  string
	IgnoredFields    []string
}

func DetailedCompare(body1, body2 string, config *VolatileConfig) (*VolatileDiff, error) {
	if config == nil {
		config = DefaultVolatileConfig()
	}

	rawDiff := body1 != body2

	normalized1Iface, err := normalizeToInterface(body1, config)
	if err != nil {
		return nil, err
	}

	normalized2Iface, err := normalizeToInterface(body2, config)
	if err != nil {
		return nil, err
	}

	normalizedEqual := reflect.DeepEqual(normalized1Iface, normalized2Iface)

	nb1b, _ := json.Marshal(normalized1Iface)
	nb2b, _ := json.Marshal(normalized2Iface)

	diff := &VolatileDiff{
		HasDiff:          rawDiff,
		VolatileOnly:     rawDiff && !normalizedEqual,
		StableFieldsDiff: !normalizedEqual,
		NormalizedBody1:  string(nb1b),
		NormalizedBody2:  string(nb2b),
		IgnoredFields:    collectIgnoredFields(body1, body2, config),
	}

	return diff, nil
}

func collectIgnoredFields(body1, body2 string, config *VolatileConfig) []string {
	var fields []string
	seen := make(map[string]bool)

	for _, body := range []string{body1, body2} {
		var data any
		if err := json.Unmarshal([]byte(body), &data); err != nil {
			continue
		}

		collectFieldNames(data, "", config, seen, &fields)
	}

	return fields
}

func collectFieldNames(data any, prefix string, config *VolatileConfig, seen map[string]bool, fields *[]string) {
	switch v := data.(type) {
	case map[string]any:
		for key, value := range v {
			fullPath := key
			if prefix != "" {
				fullPath = prefix + "." + key
			}

			if shouldIgnoreField(key, config) && !seen[fullPath] {
				seen[fullPath] = true
				*fields = append(*fields, fullPath)
			}

			collectFieldNames(value, fullPath, config, seen, fields)
		}

	case []any:
		for _, item := range v {
			collectFieldNames(item, prefix, config, seen, fields)
		}
	}
}

func ConfigFromFlags(ignoreFields, ignorePatterns []string) *VolatileConfig {
	config := DefaultVolatileConfig()

	if len(ignoreFields) > 0 {
		config.IgnoreFields = append(config.IgnoreFields, ignoreFields...)
	}

	for _, pattern := range ignorePatterns {
		if re, err := regexp.Compile(pattern); err == nil {
			config.IgnorePatterns = append(config.IgnorePatterns, re)
		}
	}

	return config
}

File: crates/replayer-cli/src/replay/volatile.rs
use std::collections::HashMap;

use regex::Regex;
use serde_json::Value;

pub struct VolatileConfig {
    pub ignore_fields: Vec<String>,
    pub ignore_patterns: Vec<Regex>,
}

pub struct VolatileDiff {
    pub has_diff: bool,
    pub volatile_only: bool,
    pub stable_fields_diff: bool,
    pub normalized_body1: String,
    pub normalized_body2: String,
    pub ignored_fields: Vec<String>,
}

pub fn default_volatile_config() -> VolatileConfig {
    VolatileConfig {
        ignore_fields: vec![
            "timestamp".to_string(),
            "createdAt".to_string(),
            "updatedAt".to_string(),
            "id".to_string(),
            "uuid".to_string(),
            "requestId".to_string(),
            "traceId".to_string(),
            "spanId".to_string(),
            "date".to_string(),
            "time".to_string(),
            "version".to_string(),
        ],
        ignore_patterns: vec![
            Regex::new(r"(?i).*_at$").unwrap(),
            Regex::new(r"(?i).*_id$").unwrap(),
            Regex::new(r"(?i).*timestamp.*").unwrap(),
            Regex::new(r"(?i).*uuid.*").unwrap(),
        ],
    }
}

pub fn config_from_flags(ignore_fields: &[String], ignore_patterns: &[String]) -> VolatileConfig {
    let mut config = default_volatile_config();

    if !ignore_fields.is_empty() {
        config.ignore_fields.extend(ignore_fields.iter().cloned());
    }

    for pattern in ignore_patterns {
        if let Ok(re) = Regex::new(pattern) {
            config.ignore_patterns.push(re);
        }
    }

    config
}

pub fn normalize_json(json_str: &str, config: &VolatileConfig) -> anyhow::Result<String> {
    let data: Value = match serde_json::from_str(json_str) {
        Ok(v) => v,
        Err(e) => return Err(anyhow::anyhow!("{}", e)),
    };

    let normalized = remove_volatile_fields(&data, config);

    match serde_json::to_string(&normalized) {
        Ok(s) => Ok(s),
        Err(e) => Err(anyhow::anyhow!("{}", e)),
    }
}

pub fn detailed_compare(
    body1: &str,
    body2: &str,
    config: &VolatileConfig,
) -> anyhow::Result<VolatileDiff> {
    let raw_diff = body1 != body2;

    let data1: Value = serde_json::from_str(body1)?;
    let data2: Value = serde_json::from_str(body2)?;

    let normalized1 = remove_volatile_fields(&data1, config);
    let normalized2 = remove_volatile_fields(&data2, config);

    let normalized_equal = normalized1 == normalized2;

    let nb1 = serde_json::to_string(&normalized1).unwrap_or_default();
    let nb2 = serde_json::to_string(&normalized2).unwrap_or_default();

    Ok(VolatileDiff {
        has_diff: raw_diff,
        volatile_only: raw_diff && normalized_equal,
        stable_fields_diff: !normalized_equal,
        normalized_body1: nb1,
        normalized_body2: nb2,
        ignored_fields: collect_ignored_fields(body1, body2, config),
    })
}

pub fn should_ignore_field(field_name: &str, config: &VolatileConfig) -> bool {
    for ignore in &config.ignore_fields {
        if field_name.eq_ignore_ascii_case(ignore) {
            return true;
        }
    }

    for pattern in &config.ignore_patterns {
        if pattern.is_match(field_name) {
            return true;
        }
    }

    false
}

pub fn collect_ignored_fields(body1: &str, body2: &str, config: &VolatileConfig) -> Vec<String> {
    let mut fields = Vec::new();
    let mut seen = HashMap::new();

    for body in [body1, body2] {
        if let Ok(data) = serde_json::from_str::<Value>(body) {
            collect_field_names(&data, "", config, &mut seen, &mut fields);
        }
    }

    fields
}

fn remove_volatile_fields(data: &Value, config: &VolatileConfig) -> Value {
    match data {
        Value::Object(map) => {
            let mut result = serde_json::Map::new();
            for (key, value) in map {
                if should_ignore_field(key, config) {
                    continue;
                }
                result.insert(key.clone(), remove_volatile_fields(value, config));
            }
            Value::Object(result)
        }
        Value::Array(arr) => Value::Array(
            arr.iter()
                .map(|item| remove_volatile_fields(item, config))
                .collect(),
        ),
        other => other.clone(),
    }
}

fn collect_field_names(
    data: &Value,
    prefix: &str,
    config: &VolatileConfig,
    seen: &mut HashMap<String, bool>,
    fields: &mut Vec<String>,
) {
    match data {
        Value::Object(map) => {
            for (key, value) in map {
                let full_path = if prefix.is_empty() {
                    key.clone()
                } else {
                    format!("{}.{}", prefix, key)
                };

                if should_ignore_field(key, config) && !seen.contains_key(&full_path) {
                    seen.insert(full_path.clone(), true);
                    fields.push(full_path.clone());
                }

                collect_field_names(value, &full_path, config, seen, fields);
            }
        }
        Value::Array(arr) => {
            for item in arr {
                collect_field_names(item, prefix, config, seen, fields);
            }
        }
        _ => {}
    }
}
